maro/simulator/scenarios/finance/sub_engines/stock/stock_business_engine.py

```diff
@@ -0,0 +1,270 @@
+                )
+                action.state = ActionState.failed
+                action.finish_tick = tick
+
+        else:
+            # order canceled
+            self._canceled_orders.remove(action.id)
+            action.state = ActionState.canceled
+            action.finish_tick = tick
+        print("result", ret)
+        return ret
+
+    def cancel_order(self, action: Action):
+        if action.id in self._pending_orders:
+            self._pending_orders.remove(action.id)
+        if action.id not in self._canceled_orders:
+            print(f'Order canceled :{action.id}')
+            self._canceled_orders.append(action.id)
+
+    def reset(self):
+        for _, reader in self._readers.items():
+            reader.reset()
+
+        self._frame.reset()
+        self._snapshots.reset()
+        self._pending_orders = []
+        self._canceled_orders = []
+
+    def _action_scope(self, action_type: ActionType, stock_index: int):
+        if action_type == ActionType.order:
+            # action_scope of stock
+            stock: Stock = self._stock_list[stock_index]
+            if self._allow_split:
+                result = (-stock.account_hold_num, stock.trade_volume)
+            else:
+                result = (
+                    max([-stock.account_hold_num, -stock.trade_volume * self._action_scope_max]),
+                    stock.trade_volume * self._action_scope_max
+                )
+            return (result, self._trader.supported_orders, self._order_mode)
+
+        elif action_type == ActionType.transfer:
+            # action_scope of account
+            result = (-self._account.remaining_money, float("inf"))
+            return result
+
+        elif action_type == ActionType.cancel_order:
+            # action_scope of pending orders
+            result = self._pending_orders
+
+            return result
+
+    def _init_frame(self):
+        self._frame = build_frame(len(self._stock_codes), self._max_tick)
+        self._snapshots = self._frame.snapshots
+        for index in range(len(self._stock_codes)):
+            stock = self._frame.stocks[index]
+            stock.set_init_state(self._stock_codes[index])
+
+    def _build_stocks(self):
+        self._stocks_dict = {}
+        self._stock_list = self._frame.stocks
+
+        for index, stock in enumerate(self._stock_list):
+            self._stocks_dict[index] = stock
+
+        self._account = self._frame.sub_account[0]
+        self._account.set_init_state(
+            currency=self._config['currency'], leverage=self._config['leverage'],
+            min_leverage_rate=self._config['min_leverage_rate'], init_money=self._config["money"]
+        )  # we only one account, so the index is 0
+
+    def _init_reader(self):
+        data_folder = self._config["data_path"]
+
+        self._stock_codes = self._config["stocks"]
+
+        # TODO: is it a good idea to open lot of file at same time?
```

Comment:
an item_picker corresponds to an opening file handler?

Reply:
file -> reader -> mult item_pickers

Reply:
So, is the ``TODO`` you list before still a problem?

Reply:
shall confirm with pingchao